network/src/noise_wrapper/handshake.rs

```diff
@@ -0,0 +1,356 @@
+        let (their_public_key, handshake_state, payload) = self
+            .0
+            .parse_client_init_message(&[], &client_init_message)
+            .map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("noise: wrong message received {}", e),
+                )
+            })?;
+
+        // make sure the public key is a validator before continuing (if we're in the validator network)
+        if let Some(trusted_peers) = trusted_peers {
+            let found = trusted_peers
+                .read()
+                .unwrap()
+                .iter()
+                .any(|(_peer_id, public_keys)| public_keys.identity_public_key == their_public_key);
+            if !found {
+                // TODO: security logging (mimoo)
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!(
+                        "noise: client connecting to us with an unknown public key: {}",
+                        their_public_key
+                    ),
+                ));
+            }
+        }
+
+        // if on a mutually authenticated network
+        if let Some(anti_replay_timestamps) = &anti_replay_timestamps {
+            // check that the payload received as the client timestamp (in seconds)
+            if payload.len() != PAYLOAD_SIZE {
+                // TODO: security logging (mimoo)
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "noise: client initiated connection without an 8-byte timestamp",
+                ));
+            }
+            let mut client_timestamp = [0u8; PAYLOAD_SIZE];
+            client_timestamp.copy_from_slice(&payload);
+            let client_timestamp = u64::from_le_bytes(client_timestamp);
+
+            // check the timestamp is not a replay
+            let mut anti_replay_timestamps = anti_replay_timestamps.write().unwrap();
+            if anti_replay_timestamps.is_replay(their_public_key, client_timestamp) {
+                // TODO: security logging the ip + blocking the ip? (mimoo)
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!(
+                        "noise: client initiated connection with a timestamp already seen before: {}",
+                        client_timestamp
+                    ),
+                ));
+            }
+
+            // store the timestamp
+            anti_replay_timestamps.store_timestamp(their_public_key, client_timestamp);
+        }
+
+        // construct the response
+        let mut rng = rand::rngs::OsRng;
+        let mut server_response = [0u8; noise::handshake_resp_msg_len(0)];
+        let session = self
+            .0
+            .respond_to_client(&mut rng, handshake_state, None, &mut server_response)
+            .map_err(|e| {
+                io::Error::new(
```

Comment:
Would an ` impl From<ThisError> for ThatError` help here (and in multiple other places)?

Reply:
yes! Good one :) looking into this

Reply:
urg, io::Error is in a different crate, so can't do this.